### PR TITLE
Updated PowerFlex/ScaleIO storage plugin to support separate (storage) network for Hosts(KVM)/Storage connection.

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR enhances the existing PowerFlex/ScaleIO storage plugin to support separate (storage) network for Hosts(KVM)/Storage connection, mainly the SDC (ScaleIo Data Client) connection.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Manually tested VM/Volume operations on PowerFlex/ScaleIO storage.

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
Hi @rohityadavcloud No UI changes in this PR, but bot detected UI changes. Can you check/fix it. Thanks.

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java [hunk 1]
```diff
@@ -78,27 +116,38 @@ public boolean hostConnect(long hostId, long poolId) {
...
         return true;
     }
 
-    private boolean isHostSdcConnected(String hostIpAddress, long poolId) {
+    private String getHostSdcId(String sdcGuid, long poolId) {
+        try {
+            ScaleIOGatewayClient client = ScaleIOGatewayClientConnectionPool.getInstance().getClient(poolId, _storagePoolDetailsDao);
+            return client.getSdcIdByGuid(sdcGuid);
+        } catch (NoSuchAlgorithmException | KeyManagementException | URISyntaxException e) {
+            s_logger.error("Failed to get host SDC id", e);
```
Comment (GutoVeronezi) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java:135]: We can add more context to the log.

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java [hunk 2]
```diff
@@ -97,6 +107,52 @@ public static final String getSystemIdForVolume(String volumeId) {
...
+            return null;
+        }
+
+        if (result.isEmpty()) {
+            LOGGER.warn("No SDC guid retrieved");
+            return null;
+        }
+
+        String guidRegEx = "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$";
+        if (!result.matches(guidRegEx)) {
```
Comment (GutoVeronezi) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:124]: We can use `UuidUtils#validateUUID` here, as it does the same validation.

**Review by DaanHoogland — APPROVED** [review]
agree with mosat of @GutoVeronezi ś comments, but generally lgtm

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java [hunk 3]
```diff
@@ -191,41 +194,59 @@ public void revokeAccess(DataObject dataObject, Host host, DataStore dataStore)
...
+            StoragePoolHostVO poolHostVO = storagePoolHostDao.findByPoolHost(poolId, hostId);
+            if (poolHostVO == null) {
+                return null;
+            }
+
+            final ScaleIOGatewayClient client = getScaleIOClient(poolId);
+            if (client.isSdcConnected(poolHostVO.getLocalPath())) {
+                return poolHostVO.getLocalPath();
+            }
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java:243]: yes, any issue with storage client is caught here & logged, and returns null.

**Comment by sureshanaparti** [comment]
@GutoVeronezi thanks for the review. addressed the review comments, please check.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]
LGTM, didn't test it.

**Comment by rohityadavcloud** [comment]
@sureshanaparti do we know anybody who has tested it?

**Comment by sureshanaparti** [comment]
> @sureshanaparti do we know anybody who has tested it?

no @rohityadavcloud, @borisstoyanov might test this, but may not be with separate host network (for storage connectivity). Testing without separate host network should be Ok, as this PR removes the dependency of the host's private address for ScaleIO operations, and uses SDC (ScaleIO Data Client) Id instead.

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java [hunk 4]
```diff
@@ -191,41 +194,59 @@ public void revokeAccess(DataObject dataObject, Host host, DataStore dataStore)
                 final VolumeVO volume = volumeDao.findById(dataObject.getId());
                 LOGGER.debug("Revoking access for PowerFlex volume: " + volume.getPath());
 
-                final ScaleIOGatewayClient client = getScaleIOClient(dataStore.getId());
-                final Sdc sdc = client.getConnectedSdcByIp(host.getPrivateIpAddress());
-                if (sdc == null) {
+                final String sdcId = getConnectedSdc(dataStore.getId(), host.getId());
+                if (StringUtils.isBlank(sdcId)) {
                     throw new CloudRuntimeException("Unable to revoke access for volume: " + dataObject.getId() + ", no Sdc connected with host ip: " + host.getPrivateIpAddress());
                 }
```
Comment (DaanHoogland) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java:200]: this bit seems to be repeated quite a few times. can we extract?

Comment (DaanHoogland) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java:200]: maybe wit an extra `boolean doAlert`

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/driver/ScaleIOPrimaryDataStoreDriver.java:243]: storage maybe you want to inform in the message that the problem is with the client and not with the MS-code?

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java [hunk 5]
```diff
@@ -69,7 +73,41 @@ public boolean hostConnect(long hostId, long poolId) {
...
+        }
+
+        if (StringUtils.isBlank(sdcId)) {
+            String msg = "Couldn't retrieve SDC details from the host: " + hostId + ", (re)install SDC and restart agent";
+            s_logger.warn(msg);
+            _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), "SDC details not found on host: " + host.getUuid(), msg);
+            return false;
+        }
+
+        if (!isHostSdcConnected(sdcId, poolId)) {
```
Comment (DaanHoogland) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java:110]: ca this bit be extracted in a few bits/modularised a bit more to improve readability?

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java [hunk 6]
```diff
@@ -53,6 +54,16 @@ public class ScaleIOUtil {
     // VOL-ID 6c33633100000009 MDM-ID 218ce1797566a00f
     // VOL-ID 6c3362a30000000a MDM-ID 218ce1797566a00f
 
+    private static final String QUERY_GUID_CMD = "drv_cfg --query_guid";
+    // Sample output for cmd: drv_cfg --query_guid:
+    // B0E3BFB8-C20B-43BF-93C8-13339E85AA50
```
Comment (DaanHoogland) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:59]: is this javadoc for the above constant?

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java [hunk 7]
```diff
@@ -53,6 +54,16 @@ public class ScaleIOUtil {
...
+
+    private static final String QUERY_MDMS_CMD = "drv_cfg --query_mdms";
+    // Sample output for cmd: drv_cfg --query_mdms:
+    // Retrieved 2 mdm(s)
+    // MDM-ID 3ef46cbf2aaf5d0f SDC ID 6b18479c00000003 INSTALLATION ID 68ab55462cbb3ae4 IPs [0]-x.x.x.x [1]-x.x.x.x
+    // MDM-ID 2e706b2740ec200f SDC ID 301b852c00000003 INSTALLATION ID 33f8662e7a5c1e6c IPs [0]-x.x.x.x [1]-x.x.x.x
```
Comment (DaanHoogland) [inline plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:65]: javadoc?

**Review by DaanHoogland — COMMENTED** [review]
clgtm generally, some readability remoarks

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java (hunk 6 above)
Comment (sureshanaparti) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:59]: > is this javadoc for the above constant?

no, just to indicate sample output for the cmd.

Comment (DaanHoogland) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:59]: Can we move this to some javadoc? it is a bit out of the blue.

Comment (sureshanaparti) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/util/ScaleIOUtil.java:59]: ok @DaanHoogland

File: plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java (hunk 5 above)
Comment (sureshanaparti) [reply plugins/storage/volume/scaleio/src/main/java/org/apache/cloudstack/storage/datastore/provider/ScaleIOHostListener.java:110]: will do this refactoring in another PR if needed. thanks.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Pull request merged**